Feature: Add support for metadata extraction from moodle URL resources

tomdickman opened the pull request.

Commit:
wip: Adding cleanup scheduled task for metadata

Commit:
refactor: refactor api to decouple from file resources

Move logic away from specific handling of file resources.
- Change contenthash to resourcehash, as some resources may have unknown
content until actually parsed by extractor
- Added a helper class to determine resource type, id, resourcehash etc.
- Added a large test base for quality control
- Updated cli script
- Make api static, in order to prevent requirement on instantiation of
an instance
- Many minor code chores
- Added URL resource type support
- Add resource validation to extractors
- Added scheduled task for URL resources

Commit:
fix: Update return value to null if no metadata

Commit:
fix: Make upgrade script field rename idempotent

Commit:
fix: cleanup file metadata task

Commit:
refactor: Move extraction processing into abstract class

Abstracted the base logic for creating metadata extraction tasks
away from resource specific tasks into a base class that can
be extended for supported resource types.
This will make it easier to implement new resource types in the
future and avoid code duplication.

Commit:
fix: process extraction task not shifting pointer correctly

When we get a chunk of resource records at each scheduled task to
create extractions for, the starting resource id should shift
by the maximum process count so the next scheduled task for this
resource knows where to start.
Fixed an issue where the id was always being reset to 0 (zero) and
starting from the beginning of table again.

Commit:
tests: Add unit tests for process extractions start id

brendanheywood commented on cli/extractmetadata.php:
```diff
@@ -27,22 +28,65 @@
+Extract metadata from a resource using the command line.
+
+Options:
+-h, --help            Print out this help
```
nicer to line these up:

```-h, --help            Print out this help
    --type (required)     The resource type (example: 'file', 'url)
    --plugin (required)   The name of the metadataextractor subplugin to use for extraction
    --id (required)       The id of the resource
-j, --json                Extract metadata as a JSON string
    --showdebugging       Print debugging statements
```

Commit:
refactor: move CRUD for metadata into metadata class itself

Commit:
fix: broken metadata extraction caused by refactor

Commit:
tests: Add word document fixture and metadata mocking

Commit:
tests: Add test coverage for metadata extraction task

Includes emergent changes identified from tests:
- Remove exception thrown from metadata populate methods
- Remove string no longer in use
- Fix logic in metadata populate method

Commit:
update: major changes, bump version

Commit:
fix: metadata variable permissions

Allow extending classes to access base variable

bump version

Commit:
chore: clean up static variable calls to utilise methods
